File: examples/rust/event_multiplexing/notifier.rs
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

use clap::Parser;
use core::time::Duration;
use iceoryx2::prelude::*;

const CYCLE_TIME: Duration = Duration::from_secs(1);

fn main() -> Result<(), Box<dyn core::error::Error>> {
    set_log_level_from_env_or(LogLevel::Info);
    let args = Args::parse();
    let node = NodeBuilder::new().create::<ipc::Service>()?;

    let event = node
        .service_builder(&args.service.as_str().try_into()?)
        .event()
        .open_or_create()?;

    let notifier = event.notifier_builder().create()?;

    while node.wait(CYCLE_TIME).is_ok() {
        notifier.notify_with_custom_event_id(EventId::new(args.event_id))?;

        println!("[service: \"{}\"] Trigger event ...", args.service);
    }

    println!("exit");

    Ok(())
}

#[derive(Parser, Debug)]
#[clap(version, about, long_about = None)]
struct Args {
    /// Defines the service to which events are emitted.
    #[clap(short, long)]
    service: String,

    /// The event id used for triggering
    #[clap(short, long, default_value_t = 0)]
    event_id: usize,
}

File: examples/python/event_multiplexing/notifier.py
# Copyright (c) 2025 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Apache Software License 2.0 which is available at
# https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
# which is available at https://opensource.org/licenses/MIT.
#
# SPDX-License-Identifier: Apache-2.0 OR MIT

"""Event Multiplexing notifier example."""

import sys

import iceoryx2 as iox2

iox2.set_log_level_from_env_or(iox2.LogLevel.Info)
node = iox2.NodeBuilder.new().create(iox2.ServiceType.Ipc)

service_name = sys.argv[1]
event_id = sys.argv[2]

event = (
    node.service_builder(iox2.ServiceName.new(service_name)).event().open_or_create()
)
max_event_id = event.static_config.event_id_max_value
notifier = event.notifier_builder().create()

COUNTER = 0
cycle_time = iox2.Duration.from_secs(1)

try:
    while True:
        node.wait(cycle_time)
        COUNTER += 1
        notifier.notify_with_custom_event_id(iox2.EventId.new(int(event_id)))

        print('[service: "', service_name, '"] Trigger event ...')
except iox2.NodeWaitFailure:
    print("exit")
